server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java

```diff
@@ -0,0 +1,261 @@
+                && (taskMaster.isHttpLoading()
+                    || serverInventoryView.isSegmentLoadedByServer(toServerName, segment))) {
+              fromServerPeon.dropSegment(segment, moveFinishCallback);
+            } else {
+              moveFinishCallback.execute(success);
+            }
+          }
+      );
+    }
+    catch (Exception e) {
+      toServer.cancelOperation(segment, SegmentState.MOVING_TO);
+      moveFinishCallback.execute(false);
+      throw new RuntimeException(e);
+    }
+
+    return true;
+  }
+
+  /**
+   * Marks the given segment as unused.
+   */
+  public boolean deleteSegment(DataSegment segment)
+  {
+    return segmentsMetadataManager.markSegmentAsUnused(segment.getId());
+  }
+
+  /**
+   * Cancels the segment operation being performed on a server if the actual
+   * state of the segment on the server matches the given currentState.
+   */
+  public boolean cancelOperation(
+      SegmentState currentState,
+      DataSegment segment,
+      ServerHolder server
+  )
+  {
+    if (!server.cancelOperation(segment, currentState)) {
+      return false;
+    }
+
+    final LoadQueuePeon peon = server.getPeon();
+    switch (currentState) {
+      case DROPPING:
+        return peon.cancelDrop(segment);
+      case MOVING_TO:
+      case LOADING:
+        return peon.cancelLoad(segment);
+      default:
+        return false;
+    }
+  }
+
+  /**
+   * Reduces the lifetimes of the segments currently being moved in all the tiers,
+   * and returns a map from tier names to the corresponding state.
+   */
+  public Map<String, TierLoadingState> reduceLifetimesOfMovingSegments()
+  {
+    return reduceLifetimesAndCreateCopy(currentlyMovingSegments);
+  }
+
+  /**
+   * Reduces the lifetimes of the segments currently being replicated in the tiers,
+   * and returns a map from tier names to the corresponding state.
+   */
+  public Map<String, TierLoadingState> reduceLifetimesOfReplicatingSegments()
+  {
+    return reduceLifetimesAndCreateCopy(currentlyReplicatingSegments);
+  }
+
+  private Map<String, TierLoadingState> reduceLifetimesAndCreateCopy(
+      Map<String, TierLoadingState> inFlightSegments
+  )
+  {
+    final Set<String> inactiveTiers = new HashSet<>();
+    inFlightSegments.forEach((tier, holder) -> {
+      if (holder.getNumProcessingSegments() == 0) {
+        inactiveTiers.add(tier);
+      }
+      holder.reduceLifetime();
+    });
+
+    // Reset state for inactive tiers
+    inactiveTiers.forEach(inFlightSegments::remove);
+
+    return Collections.unmodifiableMap(inFlightSegments);
+  }
+
+  private boolean canLoadReplica(String tier, ReplicationThrottler throttler)
+  {
+    if (!throttler.canAssignReplica(tier)) {
+      return false;
+    }
+
+    TierLoadingState tierState = currentlyReplicatingSegments.get(tier);
+    return tierState == null || tierState.getNumProcessingSegments() < throttler.getReplicationThrottleLimit();
```

Comment:
The latter part of this really "feels" like it's a throttling decision, but the throttler appears to not be making that decision?  I don't know that I have a specific suggestion, just, it caught my eye as I read the code.

Reply:
Fair point. I will clean it up.